Carousel: add opacity transition to carousel when opening

ramonjd opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

An alternative to https://github.com/Automattic/jetpack/pull/20373
Fixes https://github.com/Automattic/view-design/issues/324

## Changes proposed in this Pull Request

Hi!

This PR adds an opacity transition effect when opening a carousel slide. 🎉 

**Before** 

The slide is blank and the image appears with no fade in effect until it's loaded.

<img src="https://user-images.githubusercontent.com/6458278/127262823-740bd8ee-83cd-44bb-9f34-b3982de1c99b.gif" width="300" />


**After**

The image should fade in. 


https://user-images.githubusercontent.com/6458278/127428373-969c5f46-668e-4217-b384-83efa0e33b66.mp4


If the image hasn't loaded yet, you should see the loading spinner.

#### Does this pull request change what data or activity we track or use?
No.

## Testing instructions

Open images with cache enabled and disabled. Try throttling the network. 

Check in mobile browsers. 

Take a deep breath between tasks.

- [ ] The carousel overlay and loading spinner should appear immediately, even when loading Swiper JS for the first time
- [ ] The info UI should only appear once the carousel is initialized.
- [ ] The current slide's image should fade in both when you click on an image in the gallery, and when you reload page with a carousel hash, e.g., `/?p=7#jp-carousel-9`
- [ ] You should be able to exit the load screen by hitting the Esc key and by clicking on the close button.

Thank you!

ramonjd committed:
[not verified] Initial commit
Testing out layering opacity fade in after the initial image has loaded

ramonjd committed:
[not verified] add changelog

ramonjd committed:
[not verified] stylistic changes.

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -29,6 +29,25 @@
+
+.jp-carousel-fade-in-image {
+	animation: fadeInOpacity ease-in 0.5s 1 normal forwards;
+	-webkit-animation: fadeInOpacity ease-in 0.5s 1 normal forwards;
```
Do we still need these vendor specific extensions? I think all the browsers we support can handle the base css3 animation properties, but could be wrong.

ramonjd replied:
Good question. At first I didn't think we needed them, but I couldn't get it to work on mobile Chrome/Safari without them. 

I'll try again. Maybe I was hallucinating. I did drink a lot of Rooibos tea today.

ramonjd replied:
Yes, we do need `webkit`, but I've ditched mozilla and the other one. Thanks for picking up on that. 👍

glendaviesnz commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -29,6 +29,25 @@
+}
+
+.jp-carousel-fade-in-image {
+	animation: fadeInOpacity ease-in 0.5s 1 normal forwards;
```
I would be inclined to go with 0.2s - 200ms is quite a good timing for the eyes not to see a harsh switch, but not to notice it as a slow fade, maybe just my opinion, but I think the likes of material design recommend 150-250ms for transitions.

ramonjd replied:
How does `0.25s` grab you?

andrewserong reviewed (changes requested):
Nice idea @ramonjd! I think adding in an opacity fade is a good alternative and definitely improves the experience when opening the carousel for me, it feels much smoother.

I ran into an odd issue if I click between two images in an image gallery. In this case, the gallery was stuck in the faded out state — I'm wondering if when clicking the space between images in the gallery, if the startIndex is possibly empty, so `currentSlideObj` might be undefined?

![carousel-fade-blank-screen-sml](https://user-images.githubusercontent.com/14988353/127266467-c00277df-aa5f-4175-a87b-e6acb781b743.gif)

Also, not quite sure if it could just be something to do with my setup, too! This was testing on desktop Chrome, but let me know if there's anything else I should test.

ramonjd committed:
[not verified] Remove vendor-specific animation rules. We do need webkit however for mobile devices.
Dropping opacity fade in duration to .25 secs

Also checking for slide object. I noticed sometimes it's not there

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.css:
```diff
@@ -29,6 +29,25 @@
+}
+
+.jp-carousel-fade-in-image {
+	animation: fadeInOpacity ease-in 0.5s 1 normal forwards;
```
I updated to `.25s` but will happily drop the extra `0.05` for consistency.

ramonjd committed:
[not verified] Dropping opacity fade in duration to .2 secs

ramonjd commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1464,18 +1464,16 @@
 				};
 				jsScript.onerror = function () {
-					domUtil.hide( loader );
+					toggleLoader();
```
✅

ramonjd committed:
[not verified] Passing explicit arg to toggleLoader

ramonjd commented:
> i'm wondering if when clicking the space between images in the gallery, if the startIndex is possibly empty, so currentSlideObj might be undefined?

I also saw this! 03e99bccbaffc9ab4048cab78d8f3748cfb5a1f4 adds a check for `currentSlideObj`

andrewserong commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1602,10 +1600,50 @@
 					}, 150 );
 				}
 			} );
+			var currentSlideObj = carousel.slides[ options.startIndex ];
```
Just checked what's happening with `options.startIndex` when clicking between images in the image gallery before the carousel is opened. In that case, `options.startIndex` is `-1`. Would it work if we updated this line to something like the following to ensure we grab the first slide if one isn't specified?:

```suggestion
			var currentSlideObj = options.startIndex > 0 ? carousel.slides[ options.startIndex ] : carousel.slides[ 0 ];
```

andrewserong commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1602,10 +1600,50 @@
+				carousel.container.classList.add( 'jp-carousel-fade-in-image' );
+				domUtil.emitEvent( carousel.overlay, 'jp_carousel.afterOpen' );
+			};
+			newImage.src = currentSlideObj.attrs.src;
```
This line is throwing an error for me. I think the above suggestion to default to the first slide if an index isn't specified should fix it.

![image](https://user-images.githubusercontent.com/14988353/127268844-21776338-e670-4a39-8f8e-78a28842b665.png)

ramonjd replied:
Oh, thanks for finding that! I've applied the suggested commit. 🙇

andrewserong replied:
Awesome, thanks for that. The code changes look good to me now, and that's working. I'll just do a little more cross browser testing, then give it the ✅

andrewserong reviewed (commented):
I _think_ I found the culprit for the click-between-the-images issue, just added another couple of comments.

ramonjd committed:
[not verified] Update projects/plugins/jetpack/modules/carousel/jetpack-carousel.js

grab specific index if one isn't given

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>

andrewserong reviewed (commented):
Thanks for working on this one @ramonjd! It's testing really well for me across iOS Safari & Chrome, and on desktop Safari, Chrome, FF, and Edge.

I also tested while network throttled, and it's looking good to me with the loader before then gracefully fading into the final image (this is especially throttled so it takes quite a long time):

![carousel-fade-throttled-sml](https://user-images.githubusercontent.com/14988353/127271646-cf4ea362-0da9-479a-8bf7-f353ef985f8b.gif)

I noticed that we're currently preventing the carousel from being interacted with during its loading state, which means that you can't tap the close button or press escape to cancel out of it while loading. Instead you need to wait for either the image to finish loading, or to fail, or reload the page.

One option for addressing that could be to retain the following block so that keyboard functionality is restored before the loader finishes doing its thing:

```
			domUtil.fadeIn( carousel.overlay, function() {
				domUtil.emitEvent( carousel.overlay, 'jp_carousel.afterOpen' );
			} );
```

And then include `toggleLoader( false );` in the `jp_carousel.beforeClose` callback, around where the swiper pagination gets hidden. I _think_ that would restore letting Escape cancel out and close the carousel, but probably doesn't get the close button working.

What do you think? Is it worth the extra trouble to ensure allow users to bail out, or are we happy to keep the loader blocking? It probably only concerns slower connections. 🤔

andrewserong commented:
And of course, we could always look at that behaviour in a follow-up if you'd prefer to get this in. Otherwise, the code's looking good to me!

ramonjd commented:
> I noticed that we're currently preventing the carousel from being interacted with during its loading state, which means that you can't tap the close button or press escape to cancel out of it while loading. 

Oh true. Nice pick up again. 

I think it'd be safer to do as you suggested and allow interaction just in case. It'd be frustrating for a user not to be able to close the gallery if the image didn't load. 

I'll update and reset.

Thank you!!!!

ramonjd committed:
[not verified] This commit shows the overlay immediately, even when Swiper JS isn't loaded.
We also hide the controls in the footer until the slider is initialized.
We can also close the swiper by clicking or esc before the image has loaded.

ramonjd commented:
In a62ed04e8ebf7aacf663706aa83d91485b31830f, I reordered the HTML so that the close button floats on top of the loading spinner.

In order to do this, I had to place everything inside the overlay. This means the overlay and loading spinner appear immediately.

Previously, if Swiper JS wasn't loaded we'd only show the spinner. 

I think this is a good trade off: it display progress to the user.

andrewserong commented:
> I think this is a good trade off: it display progress to the user.

Agreed, I like this trade-off, too. And the fade in looks a little smoother now, too, that the background color is fading in as well.

This is testing well on Chrome and Safari for me, but I ran into an odd issue in Firefox and Edge that's difficult to reproduce reliably. If I click to open the carousel and hit escape rapidly a couple of times in a row (e.g. open then close, open then close, repeated a few times), I can manage to get the carousel into an odd state where it appears that it mightn't have been unloaded correctly. The result is that nothing on the page can be clicked, as the click handlers are still registered against Swiper, even though it isn't visible:

![carousel-click-error-sml](https://user-images.githubusercontent.com/14988353/127432540-63123c8b-6739-4c72-ada3-efea421c70a0.gif)

The screengrab above makes it look like the Carousel is closed, but I think it's actually just hidden:

![image](https://user-images.githubusercontent.com/14988353/127432670-2699bb3f-118c-4c8f-9e65-aa07ef76bd0b.png)

If I manually edit the element in inspect element and update the opacity to 1, I can go over and click the close button, and then I can click on the gallery again to re-open the carousel.

I'm not sure, but I'm wondering if there's a potential race condition with the event handlers for opening and closing?

andrewserong commented:
The errors aren't very visible in that screenshot, so here's the plain text version of it (basically, lines `1607` and `771` of `jetpack-carousel.js`):

```
jetpack-carousel.js?ver=10.1-alpha:1607 Uncaught TypeError: Cannot read property 'scale' of undefined
    at Swiper.<anonymous> (jetpack-carousel.js?ver=10.1-alpha:1607)
    at swiper-bundle.js:1937
    at Array.forEach (<anonymous>)
    at swiper-bundle.js:1936
    at Array.forEach (<anonymous>)
    at Swiper.emit (swiper-bundle.js:1928)
    at Swiper.onTouchEnd (swiper-bundle.js:4349)
(anonymous) @ jetpack-carousel.js?ver=10.1-alpha:1607
(anonymous) @ swiper-bundle.js:1937
(anonymous) @ swiper-bundle.js:1936
emit @ swiper-bundle.js:1928
onTouchEnd @ swiper-bundle.js:4349
84jetpack-carousel.js?ver=10.1-alpha:771 Uncaught TypeError: Cannot read property 'attrs' of undefined
    at selectSlideAtIndex (jetpack-carousel.js?ver=10.1-alpha:771)
    at Swiper.<anonymous> (jetpack-carousel.js?ver=10.1-alpha:1581)
    at swiper-bundle.js:1937
    at Array.forEach (<anonymous>)
    at swiper-bundle.js:1936
    at Array.forEach (<anonymous>)
    at Swiper.emit (swiper-bundle.js:1928)
    at Swiper.updateActiveIndex (swiper-bundle.js:2757)
    at Swiper.slideTo (swiper-bundle.js:3237)
    at Swiper.loopFix (swiper-bundle.js:3666)
```

andrewserong commented:
To closer pinpoint the issue (I think!), in the broken state, the carousel overlay is set to `opacity: 0` instead of `display: none`.

Broken state:

![image](https://user-images.githubusercontent.com/14988353/127434851-a1c44447-ebe8-4eba-bf21-d68a6e9e4897.png)

Desired state:

![image](https://user-images.githubusercontent.com/14988353/127434908-26edfb76-2d6b-4740-9940-60716dc969d2.png)

andrewserong commented on projects/plugins/jetpack/modules/carousel/jetpack-carousel.js:
```diff
@@ -1602,10 +1610,48 @@
+			var isCurrentSlideImageLoaded =
+				currentSlideImage && currentSlideImage.complete && currentSlideImage.naturalHeight !== 0;
+
+			domUtil.emitEvent( carousel.overlay, 'jp_carousel.afterOpen' );
```
I'm wondering if the issue is that this line is no longer in a callback passed to `domUtil.fadeIn`, so the escape key can be pressed before the fade in has completed?

ramonjd replied:
Thanks for going deeper here. That's a good place to start looking.

> I'm not sure, but I'm wondering if there's a potential race condition with the event handlers for opening and closing?

I'm nearly losing track of all the events we're firing off. :) I'm not I'm keen to add to the complexity, so might treat this PR as a learning exercise if we can't get it working.

ramonjd committed:
This commit removes the fadein and fadeout effect for the overlay to avoid race conditions.
We're also intializing the UI early so we can allow escaping as soon as possible.
Because of this we have to ensure we cancel out any callbacks for the JS and images to avoid running the carousel if the user click close too quickly.
This is all a test only and maybe not appropriate to go to prod without lots of testing.

ramonjd commented:
Closing this PR for now. It's looking better. If we had another week to finesse, we might polish it up and get it thoroughly tested.

More reasons and considerations for future iterations given in the issue: https://github.com/Automattic/view-design/issues/324#issuecomment-889601472
